### DISCUSSION TITLE
Push queries abort after 10 minutes of no activitiy

### DISCUSSION THREAD

**Discussion by mrt181** [tomasfabian/ksqlDB.RestApi.Client-DotNet#101] [Q&A] [answered]
I am using push queries on (source) tables that have infrequent updates.
The connections are aborted after 4 minutes.

I am using confluent cloud which does not allow to configure: [ksql.idle.connection.timeout.seconds](https://docs.confluent.io/platform/current/ksqldb/reference/server-configuration.html#ksql-idle-connection-timeout-seconds)

A workaround is mentioned here: https://github.com/confluentinc/ksql/issues/6970#issuecomment-775548708

That involves calling some endpoint like listStreams periodically.

When creating a push query I have to use `IKSqlDBContext`
```csharp
     IQbservable<MyType> qbservable = context.CreatePushQuery<MyType>();
```

This use a client under the hood to which I do not have access.
So I can not issue a periodic request using that client to keep the connection open.

Is there a way to keep push query connections alive similar to the mentioned workaround?

**Comment by tomasfabian** [comment]
You might find a configuration for `HttpClient` that can be set as follows:
```C#
public static void Configure(this IServiceCollection services, string ksqlDbUrl)
{
  services.AddDbContext<IKSqlDBContext, KSqlDBContext>(c =>
  {
    c.UseKSqlDb(ksqlDbUrl);

    c.ReplaceHttpClient<IHttpClientFactory, ksqlDB.RestApi.Client.KSql.RestApi.Http.HttpClientFactory>(httpClient => { 
        httpClient.DefaultRequestHeaders.Add("Connection", "Keep-Alive"); //Find a proper configuration option
      })
      .AddHttpMessageHandler(_ => new CustomMessageHandler());
  });
}
```
Check also https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.connectcallback?view=net-7.0#examples

**Comment by mrt181** [comment, marked as answer]
So what we did is to configure the sockets handler to send keep alive pings.

```csharp
  builder
    .Services.AddHttpClient<
      ksqlDB.RestApi.Client.KSql.RestApi.Http.IHttpClientFactory,
      ksqlDB.RestApi.Client.KSql.RestApi.Http.HttpClientFactory
    >()
    .UseSocketsHttpHandler(
      (handler, _) =>
      {
        handler.PooledConnectionLifetime = TimeSpan.FromHours(1);
        handler.KeepAlivePingPolicy = HttpKeepAlivePingPolicy.Always;
        handler.KeepAlivePingDelay = TimeSpan.FromMinutes(1);
        handler.KeepAlivePingTimeout = TimeSpan.FromMinutes(1);
      }
    )
    .SetHandlerLifetime(Timeout.InfiniteTimeSpan);
```

This is a good solution but did not work for me because the cloud infrastructure I use has a proxy managing all connections. That proxy timeouts the connection to the server and does not forward the keep alives. That proxy is not configurable.

**Comment by tomasfabian** [comment]
Thanks for sharing the solution.
